crypto/Pos/Crypto/Orphans.hs

```diff
@@ -8,51 +8,113 @@ import           Universum
+import qualified Codec.CBOR.Encoding as E
+import           Crypto.Error (CryptoFailable (..))
+import qualified Crypto.PubKey.Ed25519 as Ed25519
 import qualified Crypto.SCRAPE as Scrape
 import           Crypto.Scrypt (EncryptedPass (..))
-import qualified Crypto.Sign.Ed25519 as Ed25519
 import           Data.Aeson (FromJSON (..), ToJSON (..))
+import           Data.Aeson.Types (Parser)
+import qualified Data.ByteArray as BA
+import qualified Data.ByteString as BS
 import           Data.Hashable (Hashable)
-import           Data.SafeCopy (base, deriveSafeCopySimple)
+import           Data.SafeCopy (SafeCopy (..), base, contain,
+                     deriveSafeCopySimple, safeGet, safePut)
+import qualified Data.Text as T
 import           Serokell.Util.Base64 (JsonByteString (..))
 
 import           Pos.Binary.Class (Bi (..), decodeBinary, encodeBinary)
 
-instance Hashable Ed25519.PublicKey
-instance Hashable Ed25519.SecretKey
-instance Hashable Ed25519.Signature
 
-instance NFData Ed25519.PublicKey
-instance NFData Ed25519.SecretKey
-instance NFData Ed25519.Signature
+fromByteStringToBytes :: BS.ByteString -> BA.Bytes
+fromByteStringToBytes = BA.convert
+
+fromByteStringToScrubbedBytes :: BS.ByteString -> BA.ScrubbedBytes
+fromByteStringToScrubbedBytes = BA.convert
+
+toByteString :: (BA.ByteArrayAccess bin) => bin -> BS.ByteString
+toByteString = BA.convert
+
+
+instance Hashable Ed25519.PublicKey where
+    hashWithSalt salt pk = hashWithSalt salt $ toByteString pk
+
+instance Hashable Ed25519.SecretKey where
+    hashWithSalt salt pk = hashWithSalt salt $ toByteString pk
+
+instance Hashable Ed25519.Signature where
+    hashWithSalt salt pk = hashWithSalt salt $ toByteString pk
+
+
+instance Ord Ed25519.PublicKey where
+    compare x1 x2 = compare (toByteString x1) (toByteString x2)
+
+instance Ord Ed25519.SecretKey where
+    compare x1 x2 = compare (toByteString x1) (toByteString x2)
+
+instance Ord Ed25519.Signature where
+    compare x1 x2 = compare (toByteString x1) (toByteString x2)
+
+
+
+instance SafeCopy BA.Bytes where
+    putCopy s = contain $ safePut (toByteString s)
+    getCopy = contain $ fromByteStringToBytes <$> safeGet
+
+instance SafeCopy BA.ScrubbedBytes where
+    putCopy s = contain $ safePut (toByteString s)
+    getCopy = contain $ fromByteStringToScrubbedBytes <$> safeGet
+
 
 deriveSafeCopySimple 0 'base ''Ed25519.PublicKey
 deriveSafeCopySimple 0 'base ''Ed25519.SecretKey
 deriveSafeCopySimple 0 'base ''Ed25519.Signature
 
+
+fromCryptoToDecoder :: T.Text -> CryptoFailable a -> Decoder s a
+fromCryptoToDecoder item (CryptoFailed e) = fail $ T.unpack $ "Pos.Crypto.Orphan.decode " <> item <> " failed because " <> show e
+fromCryptoToDecoder _ (CryptoPassed r) = return r
+
+fromCryptoToAeson :: T.Text -> CryptoFailable a -> Parser a
+fromCryptoToAeson item (CryptoFailed e) = fail $ T.unpack $ "Pos.Crypto.Orphan.parseJSON "  <> item <> " failed because " <> show e
+fromCryptoToAeson _ (CryptoPassed r) = return r
```

Comment:
Very minor but these functions are roughly the same if we consider the last argument as a `MonadFail` ? 

```hs
fromCryptoFailable :: MonadFail m :: T.Text -> CryptoFailable a -> m a
```

?

Reply:
Absolutely!!! Made the change